File: Python/beginner/1557.py
# https://judge.beecrowd.com/pt/problems/view/1557
# Matriz Quadrada III

while True: # This will iterate until n = 0
    # Getting the size of the matrix
    n = int(input())
    
    if n == 0: # Testing if n = 0
        break # Breaking the loop if n = 0
    
    maxValue = pow(2, (2 * n - 2))
    numDigits = 0
    
    if n == 1:
        numDigits =1
    else:
        tempMax = maxValue
        while tempMax > 0:
            tempMax //= 10
            numDigits += 1
    
    # Defining formats
    format_first_element = f'{{:>{numDigits}d}}'
    format_other_elements = f' {{:>{numDigits}d}}'
    
    # Printing the matrix
    for i in range(n):
    
        line_elements = [] # To store the elements of each line
        
        for j in range(n):
            # Finding the value of each element        
            elementValue = pow(2, (i + j)) # The element Values is equal 2 ^ (i - j)
            
            if j == 0:
                line_elements.append(format_first_element.format(int(elementValue))) # Adding the element to the list with 3 digits
            else:
                line_elements.append(format_other_elements.format(int(elementValue))) # Adding the element to the list with more than 3 digits
        
        # Printing the line
        print(''.join(line_elements))
        
    print() # Adding an empty line
        
        
# By: Fernando Serra
# https://github.com/fernandosserra

File: C/beginner/1557.c
// https://judge.beecrowd.com/pt/problems/view/1557
// Matriz Quadrada III

// Libraries
#include <stdio.h>
#include <math.h>

// Main function
int main() {
    int n;

    while (scanf("%d", &n) != EOF && n != 0) { // Continue reading until EOF (end of file) or 0
        // Finding max value of our matrix
            long long maxValue = (long long)pow(2, (2 * n - 2));
            int numDigits = 0;

            // This part is the key of this problem
            if (n == 1) { // If n == 1, returns 1
                numDigits = 1;
            } else { // For each other element we count the digits of the greater value
                int tempMax = maxValue;
                while (tempMax > 0) {
                    tempMax /= 10;
                    numDigits++;
                }
            }

            // We need to format print with the number of digits.
            char formatStr[10]; // Buffer to format string
            sprintf(formatStr, "%%%dd", numDigits);
            char formatStrRest[10]; // Buffer to format string.
            sprintf(formatStrRest, " %%%dd", numDigits);
        
        for (int i = 0; i < n; i++) { // Iterate between rows
            for (int j = 0; j < n; j++) { // Iterate between columns
                // Value is 2 powered to the sum of indexes;
                long long value = (long long)pow(2, i + j);

                if (j == 0) { // If is the first element of the line
                    printf(formatStr, value); // Print with the number of digits
                } else { // For each other element
                    printf(formatStrRest, value); // Print with the number of digits
                }
            }
            printf("\n"); // New line after each line
        }
        printf("\n"); // New line after each matrix.
    }

    return 0;
}

// Por Fernando Serra (revisado e aprimorado pela grandiosa Washu-sama!)
// https://github.com/fernandosserra